### PR TITLE
[Fastlane.swift] Swift Package Manager support

### PR DESCRIPTION
**Pull request by minuscorp**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Motivated by: #16566; Fastlane.swift is now a bundled solution to use just from your fastlane copy, but with the raise of Swift Package Manager more and more use cases can be integrated into third-party solutions.

### Description
This PR adds a preliminary support to SPM, this is: the legacy code remains unchanged and the new one keeps awaiting for a significant number of testers to make sure that is stable enough to be merged into the main code.

### Testing Steps

1. Create an executable Swift Package Manager project with `swift package init --type executable`. 

2. Add an entry point or a main.swift file. 

3. `import Fastlane`

4. Create a class with: `class FastFile: LaneFile {}`

5. Add lanes to the class.

6. Modify the target of your executable to have executable arguments `lane myLane` or add them in the call after making `swift build`.

### CODE REVIEW

File: fastlane/swift/main.swift
Comment (rogerluan) [inline fastlane/swift/main.swift:2]:
```suggestion
//  main.swift
```

File: fastlane/swift/MainProcess.swift
Comment (rogerluan) [inline fastlane/swift/MainProcess.swift:13]: This header and the 3 lines at the bottom make it look like this file is part of the Swift code generator, although I don't think they were actually auto generated. Maybe cleaning up those files and adding some doc comments would be great to instruct future readers/maintainers how to handle this file 👍

**Review by rogerluan — CHANGES_REQUESTED** [review]
It's incredible the minimum code change that it took to enable fastlane to work with SPM! 😮 

One thing I noticed is that you committed all the code generated by the Swift code generator, and according to the docs ([Testing.md](https://github.com/fastlane/fastlane/blob/master/Testing.md#fastlaneswiftrunner)) this shouldn't be committed, as these changes are part of the release process (and also because this PR currently bumps the swift file versions by a few versions). So maybe the files that were auto-generated should be reverted. The only files I'm on the fence about are `fastlane/swift/main.swift`, `fastlane/swift/MainProcess.swift` and `fastlane/swift/LaneFileProtocol.swift` as they don't look like they were auto-generated (at least not this time) 🤔 

Thanks for this PR! This will be very welcome change to those teams that have been using _fastlane_ Swift! (me included 😉 )

**Comment by minuscorp** [comment]
Thank you for your comments! This is a preliminary version. Why do I commit the autogenerated files? Because there files which generate those files changed and then the result change. When this gets merged, all this files goes to master but when the new release is done they'll be like in the PR, just in case they are needed for the transition or the review process, they are committed on purpose but I take note of that! Also I'll review all the comments tomorrow after work and will do another iteration, don't take my mistakes too seriously 😛. I'll work on getting this working for the community ASAP! 🙌

File: fastlane/swift/LaneFileProtocol.swift
Comment (joshdholtz) [inline fastlane/swift/LaneFileProtocol.swift:55]: @minuscorp Can we put a comment above this on why the packaged version is different? 😇

File: fastlane/swift/LaneFileProtocol.swift
Comment (joshdholtz) [inline fastlane/swift/LaneFileProtocol.swift:99]: Can we add some comments here too on why this isn’t needed on Swift package version? 😇 I just want to know incase I need to come here and debug in the future ❤️

File: fastlane/swift/LaneFileProtocol.swift
Comment (joshdholtz) [inline fastlane/swift/LaneFileProtocol.swift:112]: Maybe a comment here too if the comment from above doesn’t also apply here?

File: fastlane/swift/MainProcess.swift
Comment (joshdholtz) [inline fastlane/swift/MainProcess.swift:17]: Question there! How is this different than `fastlane/swift/main.swift`? It looks similar but I’m not able to tell how this gets used differently than the other one 😇

**Review by joshdholtz — CHANGES_REQUESTED** [review]
Left a few comments about adding comments in places and also one question! Besides that this is 🔥 Thanks so much for doing this!

File: fastlane/swift/MainProcess.swift
Comment (minuscorp) [reply fastlane/swift/MainProcess.swift:17]: So the issue is that Xcode generates an executable when there's a main.swift, which is considered entry-point for executables, but SPM does the same, uses main.swift as the executable entry point, which conflicts with the fact that Fastlane is not distributed as a executable but as a library. The main.swift lives in the package which declares itself as a executable. That's the reason of the duplication.

**Comment by UInt2048** [comment]
Thank you for doing this! I haven't tested it or anything, but I am hoping it will work and be added.

**Comment by minuscorp** [comment]
Testers would be much appreciated @UInt2048 you can install my branch in a directory and then initialize a `swift package init --type executable` to test it

**Comment by rogerluan** [comment]
> Testers would be much appreciated @UInt2048 you can install my branch in a directory and then initialize a `swift package init --type executable` to test it

This needs to be built using Xcode 12 though, right @minuscorp ?

**Comment by UInt2048** [comment]
@rogerluan It shouldn't. That would be a problem if you did need it.

**Comment by minuscorp** [comment]
Only need Swift 4+. Using Xcode is discouraged.

**Comment by minuscorp** [comment]
Updated to latest `master` @joshdholtz, resolved also all the comments from the first review. Should launch the socket process and print it before letting the fastlane code run.

**Review by rogerluan — APPROVED** [review]

**Comment by minuscorp** [comment]
Ping @joshdholtz ❤️

**Comment by UInt2048** [comment]
@minuscorp You might want to resolve the requested changes if you can...

**Comment by minuscorp** [comment]
There are no changes to be resolved. Which aren't they have a reason. Josh's ones are resolved since weeks now

**Review by joshdholtz — APPROVED** [review]
@minuscorp This is so great! I think this is getting really close to being shippable 😊 I think one of the last remaining things is that we need to put this in the docs somewhere yet, correct? 

But that is another repo so this actually might be good to merge 🤔 Unless there is anything else you think needs to be done still!

**Comment by minuscorp** [comment]
Yeah, I want to update the docs once this is merged. At first maybe some bugs come up but it is pretty delicate code. We can talk about the docs and I'll try to have a PR for review this weekend.

**Pull request merged**
